src/init/starship.osh

```diff
@@ -0,0 +1,119 @@
+function starship_precmd() {
+    # Save the status, because commands in this pipeline will change $?
+    STARSHIP_CMD_STATUS=$? STARSHIP_PIPE_STATUS=("${PIPESTATUS[@]}")
+    if [[ -n "${BP_PIPESTATUS-}" ]] && [[ "${#BP_PIPESTATUS[@]}" -gt 0 ]]; then
+        STARSHIP_PIPE_STATUS=("${BP_PIPESTATUS[@]}")
+    fi
+
+    # Due to a bug in certain Bash versions, any external process launched
+    # inside $PROMPT_COMMAND will be reported by `jobs` as a background job:
+    #
+    #   [1]  42135 Done                    /bin/echo
+    #
+    # This is a workaround - we run `jobs` once to clear out any completed jobs
+    # first, and then we run it again and count the number of jobs.
+    #
+    # More context: https://github.com/starship/starship/issues/5159
+    # Original bug: https://lists.gnu.org/archive/html/bug-bash/2022-07/msg00117.html
+    jobs &>/dev/null
+
+    local job NUM_JOBS=0 IFS=$' \t\n'
+    # Evaluate the number of jobs before running the preserved prompt command, so that tools
+    # like z/autojump, which background certain jobs, do not cause spurious background jobs
+    # to be displayed by starship. Also avoids forking to run `wc`, slightly improving perf.
+    for job in $(jobs -p); do [[ $job ]] && ((NUM_JOBS++)); done
+
+    # Run the bash precmd function, if it's set. If not set, evaluates to no-op
+    "${starship_precmd_user_func-:}"
+
+    # Set $? to the preserved value before running additional parts of the prompt
+    # command pipeline, which may rely on it.
+    _starship_set_return "$STARSHIP_CMD_STATUS"
+
+    if [[ -n "${STARSHIP_PROMPT_COMMAND-}" ]]; then
+        eval "$STARSHIP_PROMPT_COMMAND"
+    fi
+
+    local -a ARGS=(--terminal-width="${COLUMNS}" --status="${STARSHIP_CMD_STATUS}" --pipestatus="${STARSHIP_PIPE_STATUS[*]}" --jobs="${NUM_JOBS}" --shlvl="${SHLVL}")
+    # Prepare the timer data, if needed.
+    if [[ -n "${STARSHIP_START_TIME-}" ]]; then
+        STARSHIP_END_TIME=$(::STARSHIP:: time)
+        STARSHIP_DURATION=$((STARSHIP_END_TIME - STARSHIP_START_TIME))
+        ARGS+=( --cmd-duration="${STARSHIP_DURATION}")
+        STARSHIP_START_TIME=""
+    fi
+    PS1="$(::STARSHIP:: prompt "${ARGS[@]}")"
+    STARSHIP_PREEXEC_READY=true  # Signal that we can safely restart the timer
+}
+
+# We want to avoid destroying an existing DEBUG hook. If we detect one, create
+# a new function that runs both the existing function AND our function, then
+# re-trap DEBUG to use this new function. This prevents a trap clobber.
+eval "STARSHIP_DEBUG_TRAP=($(trap -p DEBUG))"
+STARSHIP_DEBUG_TRAP=("${STARSHIP_DEBUG_TRAP[2]}")
+if [[ -z "$STARSHIP_DEBUG_TRAP" ]]; then
+    trap 'starship_preexec "$_"' DEBUG
+elif [[ "$STARSHIP_DEBUG_TRAP" != 'starship_preexec "$_"' && "$STARSHIP_DEBUG_TRAP" != 'starship_preexec_all "$_"' ]]; then
+    starship_preexec_all() {
+	local PREV_LAST_ARG=$1 ; eval -- "$STARSHIP_DEBUG_TRAP"; starship_preexec; : "$PREV_LAST_ARG";
+    }
+    trap 'starship_preexec_all "$_"' DEBUG
+fi
```

Comment:
`osh` should support `PS0` which should preferably be used.

Reply:
I've [opened the discussion on oils](https://oilshell.zulipchat.com/#narrow/channel/121540-oil-discuss/topic/Any.20interest.20in.20supporting.20PS0.3F/with/527294003), but for now will keep using DEBUG